### ISSUE TITLE
Polyline Intersects Polygon Errors

### ISSUE THREAD

**Issue by katterfelto** [wfletcher/EasyGIS.NET#50]
I have added the ability to allow the user to draw a polygon with the intention that any objects intersecting the shape are marked as selected. The resulting selection is not very accurate. I have tracked the issue down to the NativeMethods.PolyLinePolygonIntersect call, but it is years since I worked with C++ so wondered if someone else could look at this.

My polygon is a bowtie shape which probably contributes the problem, but non crossing polygons also cause me intersection issues.

I have a NUnit based unit test project to highlight the problem which I have attached. If you execute these tests you will see that the individual points of the line fall within the polygon but when the polyline is assessed it does not.

[UnitTests.zip](https://github.com/wfletcher/EasyGIS.NET/files/8364719/UnitTests.zip)

**Comment by wfletcher** [comment]
The native method is a wrapper around the c++ boost geometry library. More likely to be on the c# side so I’ll look into it. Thanks for attaching the unit tests - will be quite useful and something the library has needed for awhile.

**Comment by katterfelto** [comment]
I have done some more testing.
- If I re-arrange the polygon points to form a rectangular shape rather than a bow-tie the line is detected as intersecting the polygon.
- Extending the line at the start outside the polygon still fails
- Extending the line at the end outside the polygon detects the intersection

I have attached a update to my test class

[IntersectPolygonTests.zip](https://github.com/wfletcher/EasyGIS.NET/files/8371707/IntersectPolygonTests.zip)

**Comment by wfletcher** [comment]
I have looked into this and I believe that I have found the cause of the issue. Technically your polygon in your unit test is not closed - by this I mean that the last point is not the same as the first point.

If you update the polygon then all the tests pass.

 polygon = new PointD[]
            {
                new PointD(-1027.57124731682, 260.152920724245),
                new PointD(-984.545440236484, 710.791636986736),
                new PointD(-548.626079027792, 332.617437912183),
                new PointD(-501.071239623208, 593.036796556336),
                **new PointD(-1027.57124731682, 260.152920724245)** //close the polygon so the end point is the same as the first
            };

The GeometryAlgorithms.PointInPolygon function passes as this is written in C# and I think that the algorithm works even if the polygon is not "closed."

The NativeMethods.PolyLinePolygonIntersect is written in C++ and is using the Boost geometry libraries. I suspect that this library is expecting the polygon to be properly formed and "closed".

I actually think the unit tests are a good feature and I am going to add the project to the main solution.
